File: day13/day13.rb
def parse_file
	File.open("day13Input.txt") do |f|
		patterns = Array.new
		pattern = Array.new
		f.each_line(chomp: true) do |line|
			if line == ""
				patterns.push(pattern)
				pattern = Array.new
			else
				pattern.push(line)
			end
		end
		patterns.push(pattern)
		return patterns
	end
end

def check_for_smudge_vert(mirrorLeft, rightSide)
	diffStr = ""
	line1 = mirrorLeft.size < rightSide.size ? mirrorLeft : rightSide
	line2 = mirrorLeft.size < rightSide.size ? rightSide : mirrorLeft
	line1.chars.each_with_index do |char, i|
		if char == line2[i]
			diffStr += " "
		else
			diffStr += "X"
		end
	end
	return diffStr.count("X") == 1
end

def check_for_smudge_horiz(line1, line2)
	diffStr = ""
	line1.chars.each_with_index do |char, i|
		if char == line2[i]
			diffStr += " "
		else
			diffStr += "X"
		end
	end
	return diffStr.count("X") == 1
end

def get_reflections(smudges = false)
	patterns = parse_file
	patternSummary = 0
	patterns.each do |pattern|
		possibleVertReflection = false
		## Check first line for vertical line of reflection
		addToSummary = 0
		for i in 0...pattern[0].size-1
			leftSide = pattern[0][..i]
			rightSide = pattern[0][i+1..]
			mirrorLeft = leftSide.reverse
			if mirrorLeft.start_with?(rightSide) or rightSide.start_with?(mirrorLeft) or (smudges and check_for_smudge_vert(mirrorLeft, rightSide))
				## Reflection found, check for all line
				possibleVertReflection = true
				smudgeUsed = (not mirrorLeft.start_with?(rightSide) and not rightSide.start_with?(mirrorLeft))
				pattern[1..].each do |line|
					leftSide = line[..i]
					rightSide = line[i+1..]
					mirrorLeft = leftSide.reverse
					unless mirrorLeft.start_with?(rightSide) or rightSide.start_with?(mirrorLeft)
						if smudges and not smudgeUsed and check_for_smudge_vert(mirrorLeft, rightSide)
							smudgeUsed = true
						else
							possibleVertReflection = false
							break
						end
					end
				end
				if possibleVertReflection
					if not smudges or smudgeUsed
						addToSummary = i+1
						patternSummary += i+1
						break
					end
				end
			end
		end
		## If no vertical line of reflection, check for horizontal
		for i in 0...pattern.size-1
			possibleHorizReflection = false
			if pattern[i] == pattern[i+1] or (smudges and check_for_smudge_horiz(pattern[i], pattern[i+1]))
				possibleHorizReflection = true
				smudgeUsed = pattern[i] != pattern[i+1]
				a = i-1
				b = i+2
				while a >= 0 and b < pattern.size
					if pattern[a] != pattern[b] 
						if smudges and not smudgeUsed and check_for_smudge_horiz(pattern[a], pattern[b])
							smudgeUsed = true
						else
							possibleHorizReflection = false
							break
						end
					end
					a -= 1
					b += 1
				end
				if possibleHorizReflection
					if not smudges or smudgeUsed
						addToSummary = 100 * (i + 1)
						patternSummary += 100 * (i + 1)
						break
					end
				end
			end
		end
		if smudges
		end
	end
	return patternSummary
end

p get_reflections
p get_reflections(true)

File: day13/day13.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"slices"
)

func print_error(err error) {
	if err != nil {
		fmt.Println(err)
	}
}

func parse_file() [][]string {
	file, err := os.Open("day13Input.txt")
	print_error(err)
	defer file.Close()

	patterns := [][]string{}
	currentPattern := []string{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			currentPattern = append(currentPattern, line)
		} else {
			patterns = append(patterns, currentPattern)
			currentPattern = []string{}
		}
	}
	patterns = append(patterns, currentPattern)
	return patterns
}

func check_for_smudge(aStr, bStr string) bool {
	minLength := len(aStr)
	if len(bStr) < minLength {
		minLength = len(bStr)
	}
	diffCount := 0
	for i := 0; i < minLength; i++ {
		if aStr[i] != bStr[i] {
			diffCount += 1
		}
	}
	return (diffCount == 1)
}

func get_reflections(smudges bool) {
	patterns := parse_file()
	patternSummary := 0
	for _, pattern := range patterns {
		// Check for vertical line of reflection
		possibleVertReflection := false
		line := pattern[0]
		for i := 1; i < len(line); i++ {
			smudgeUsed := false
			rightSide := line[i:]
			leftSide := line[:i]
			leftSideRunes := []rune(leftSide)
			slices.Reverse(leftSideRunes)
			mirrorLeftSide := string(leftSideRunes)
			// If either part of the string is equal to the start of the other, possible reflection found, check all other lines
			if strings.HasPrefix(mirrorLeftSide, rightSide) || strings.HasPrefix(rightSide, mirrorLeftSide) || (smudges && check_for_smudge(mirrorLeftSide, rightSide)) {
				possibleVertReflection = true
				smudgeUsed = (!strings.HasPrefix(mirrorLeftSide, rightSide) && !strings.HasPrefix(rightSide, mirrorLeftSide))
				for _, reflLine := range pattern[1:] {
					checkRight := reflLine[i:]
					checkLeft := reflLine[:i]
					checkLeftRunes := []rune(checkLeft)
					slices.Reverse(checkLeftRunes)
					checkMirrorLeft := string(checkLeftRunes)
					if !strings.HasPrefix(checkMirrorLeft, checkRight) && !strings.HasPrefix(checkRight, checkMirrorLeft) {
						if smudges && !smudgeUsed && check_for_smudge(checkMirrorLeft, checkRight) {
							smudgeUsed = true
						} else {
							possibleVertReflection = false
							break
						}
					}
				}
			}
			if possibleVertReflection {
				if !smudges || smudgeUsed {
					patternSummary += i
					break
				}
			}
		}
		// Check for horizontal line of reflection
		for i := 0; i < len(pattern) - 1; i++ {
			possibleHorizReflection := true
			smudgeUsed := false
			for a, b := i, i+1; a >= 0 && b < len(pattern); a, b = a-1, b+1 {
				aLine := pattern[a]
				bLine := pattern[b]
				if aLine != bLine {
					if smudges && !smudgeUsed && check_for_smudge(aLine, bLine) {
						smudgeUsed = true
					} else {
						possibleHorizReflection = false
						break
					}
				}
			}
			if possibleHorizReflection {
				if !smudges || smudgeUsed {
					patternSummary += (100*(i+1))
					break
				}
			}
		}
	}
	fmt.Println(patternSummary)
}

func main() {
	get_reflections(false)
	get_reflections(true)
}